packages/analytics-js-plugins/src/deviceModeTransformation/index.ts

```diff
@@ -17,25 +39,159 @@ const DeviceModeTransformation = (): ExtensionPlugin => ({
     state.plugins.loadedPlugins.value = [...state.plugins.loadedPlugins.value, pluginName];
   },
   transformEvent: {
+    init(
+      state: ApplicationState,
+      pluginsManager: IPluginsManager,
+      httpClient: IHttpClient,
+      storeManager: IStoreManager,
+      errorHandler?: IErrorHandler,
+      logger?: ILogger,
+    ) {
+      const writeKey = state.lifecycle.writeKey.value as string;
+      httpClient.setAuthHeader(writeKey);
+
+      const eventsQueue = new RetryQueue(
+        // adding write key to the queue name to avoid conflicts
+        `${QUEUE_NAME}_${writeKey}`,
+        DEFAULT_TRANSFORMATION_QUEUE_OPTIONS,
+        (item: TransformationQueueItemData, done: DoneCallback) => {
+          httpClient.getAsyncData({
+            url: `${state.lifecycle.dataPlaneUrl.value}/transform`,
+            options: {
+              method: 'POST',
+              data: getDMTDeliveryPayload(item.payload) as string,
+              sendRawData: true,
+            },
+            isRawResponse: true,
+            timeout: REQUEST_TIMEOUT_MS,
+            callback: (result, details) => {
+              // null means item will not be requeued
+              const queueErrResp = isErrRetryable(details) ? details : null;
+
+              pluginsManager.invokeSingle(
+                'transformEvent.sendDataToDestination',
+                state,
+                pluginsManager,
+                result,
+                details?.xhr?.status,
+                item.event,
+                errorHandler,
+                logger,
+              );
+              done(queueErrResp, result);
+            },
+          });
+        },
+        storeManager,
+        MEMORY_STORAGE,
+      );
+
+      return eventsQueue;
+    },
+
     enqueue(
+      state: ApplicationState,
+      eventsQueue: IQueue,
+      event: RudderEvent,
+      destinations: Destination[],
+    ) {
+      const destinationIds = destinations.map(d => d.id);
+      const payload = createPayload(event, destinationIds, state.session.authToken.value);
+      eventsQueue.addItem({ event, payload } as TransformationQueueItemData);
```

Comment:
You can enqueue the event, dest IDs, and token to the queue and prepare the payload while consuming the queue item in the callback.

Reply:
Done